engine/class_modules/priest/sc_priest_shadow.cpp

```diff
@@ -1002,27 +1001,56 @@ struct devouring_plague_t final : public priest_spell_t
     }
   }
 
-  // TODO override refresh duration to match in-game refresh
+  timespan_t calculate_dot_refresh_duration( const dot_t* d, timespan_t duration ) const override
+  {
+    // when you refresh, you lose the partial tick
+    return duration + d->time_to_next_tick();
+  }
 
   void snapshot_state( action_state_t* s, result_amount_type rt ) override
   {
     priest_spell_t::snapshot_state( s, rt );
 
     double multiplier = 1.0;
-    dot_t* dot = get_dot( s->target );
+    dot_t* dot        = get_dot( s->target );
     if ( dot->is_ticking() )
     {
       action_state_t* old_s = dot->state;
       action_state_t* new_s = s;
 
       timespan_t time_to_tick = dot->time_to_next_tick();
-      timespan_t old_remains = dot->remains();
-      timespan_t new_remains = calculate_dot_refresh_duration( dot, composite_dot_duration( new_s ) );
-      timespan_t old_tick = tick_time( old_s );
-      timespan_t new_tick = tick_time( new_s );
-      double old_multiplier = cast_state( old_s )->rolling_multiplier;
+      timespan_t old_remains  = dot->remains();
+      timespan_t new_remains  = calculate_dot_refresh_duration( dot, composite_dot_duration( new_s ) );
+      timespan_t old_tick     = tick_time( old_s );
+      timespan_t new_tick     = tick_time( new_s );
+      double old_multiplier   = cast_state( old_s )->rolling_multiplier;
+
+      // old_remains * old_multiplier / new_remains * X
+      int num_full_ticks      = as<int>( std::floor( ( old_remains - time_to_tick ) / old_tick ) );
+      double tick_coefficient = dot->current_action->spell_tick_power_coefficient( old_s ) * old_multiplier;
+
+      // find number of ticks in new DP
+      double new_num_ticks        = new_remains / new_tick;
+      double new_tick_coefficient = dot->current_action->spell_tick_power_coefficient( new_s );
+
+      if ( sim->debug )
+      {
+        sim->print_debug(
+            "Devouring Plague calculations - num_full_ticks: {}, tick_coefficient: {}, new_num_ticks: {}, "
+            "new_tick_coefficient: {}",
+            num_full_ticks, tick_coefficient, new_num_ticks, new_tick_coefficient );
+      }
+
+      // figure out the increase for each new tick of DP
+      multiplier = 1 + ( ( ( num_full_ticks * tick_coefficient ) / new_num_ticks ) / new_tick_coefficient );
```

Comment:
You might want to split this up into 3 steps to make it easier to comprehend and maintain.
First the ratio from how old ticks are distributed onto the new ones.
Then how coefficient ratio affects that.
And just when assigning adding the 1 to ensure the new dot also gets its base multiplier.